data/src/test/java/org/apache/iceberg/data/DataTestHelpers.java

```diff
@@ -88,6 +90,16 @@ private static void assertEquals(Type type, Object expected, Object actual) {
             "Primitive value should be equal to expected for type " + type, expected, actual);
         break;
       case FIXED:
+        // For fixed type, Iceberg actually represent value as Bytebuffer,
+        // but since RandomGenericData generates bytearray data for fixed type
+        // for it to be written to Avro, we add a conversion here to make the
+        // equality comparison consistent using bytearray
+        if (expected instanceof ByteBuffer) {
+          expected = ByteBuffers.toByteArray((ByteBuffer) expected);
+        }
+        if (actual instanceof ByteBuffer) {
+          actual = ByteBuffers.toByteArray((ByteBuffer) actual);
```

Comment:
This looks suspicious. Are you sure that there is a mismatch in the readers?

Reply:
Yes, in `Type.java`, fixed type is associated with `ByteBuffe`:
```
    FIXED(ByteBuffer.class),
```
So theoretically, fixed type data should always be `BtyeBuffer` type instead of byte array.
And here the problem is actually the `RandomGenericData` generated the `byte[]` data for fixed type ([I tried to change that one](https://github.com/apache/iceberg/blob/8271791061e72dbf554028e477746e89fca9ec02/data/src/test/java/org/apache/iceberg/data/RandomGenericData.java#L236) but it resulted in more tests failures because there are some existing avro writers depending on it, and Avro can only write fixed type data represented by `byte[]`).

So I instead added the conversion in this assertion helper class to convert the ByteBuffer back to `byte[]` for compatible comparison.

Reply:
For the Iceberg generic in-memory representation, `byte[]` is correct. The reader should be producing `byte[]` and not `ByteBuffer`. The Iceberg type's parameter is for Iceberg internals, not the generic representation.

Reply:
In the latest patch, this check is removed, and `SingleValueParser` moved to `byte[]`. However, it seems that `Literals` still uses `ByteBuffer` for `FIXED` type, so the discrepancy showed up again in `SingleValueParser`, where it needs to handle both `byte[]` and `ByteBuffer` inputs. There is a more detailed comment on this in `SingleValueParser`.

Reply:
Also to clarify where the dependency on `byte[]` is coming from, it actually comes from `DataTestHelpers`. We can see [here](https://github.com/apache/iceberg/blob/d04efee702fcdcdbe3659c12f7442f5000aa246a/data/src/test/java/org/apache/iceberg/data/DataTestHelpers.java#L90) that the `FIXED` case fails the assertion if the data type is not `byte[]`. Hence, there is a conflict between `DataTestHelpers` which uses `byte[]` and `Literals` which uses `ByteBuffer`. Such a gap is bridged in `SingleValueParser#toJson` in the current PR.

Reply:
@wmoustafa what is the failure?

Reply:
If we keep `SingleValueParser` and `DataTestHelpers` without any change, the error is:
```
org.apache.iceberg.data.avro.TestReadDefaultValues > writeAndValidate FAILED
    java.lang.AssertionError: [Expected should be a byte[]] 
    Expecting actual:
      java.nio.HeapByteBuffer[pos=0 lim=2 cap=2]
    to be an instance of:
      byte[]
    but was instance of:
      java.nio.HeapByteBuffer
        at org.apache.iceberg.data.DataTestHelpers.assertEquals(DataTestHelpers.java:93)
        at org.apache.iceberg.data.DataTestHelpers.assertEquals(DataTestHelpers.java:39)
        at org.apache.iceberg.data.avro.TestReadDefaultValues.writeAndValidate(TestReadDefaultValues.java:164)
```

Reply:
Okay, I think this does indicate a bug.

You're reading into Iceberg generics, so the values should all match the classes created by Iceberg generics. Those are not the same as the internal classes used by Iceberg because they are intended to be friendlier. For example, Iceberg users days from the unix epoch to represent a date, but generics will return Java's `LocalDate`.

Here's a diff that fixes your tests by using the correct types for the expected values:
```diff
diff --git a/data/src/test/java/org/apache/iceberg/data/avro/TestReadDefaultValues.java b/data/src/test/java/org/apache/iceberg/data/avro/TestReadDefaultValues.java
index 0f98d2e571..21f1805877 100644
--- a/data/src/test/java/org/apache/iceberg/data/avro/TestReadDefaultValues.java
+++ b/data/src/test/java/org/apache/iceberg/data/avro/TestReadDefaultValues.java
@@ -23,6 +23,7 @@ import static org.apache.iceberg.types.Types.NestedField.required;
 
 import java.io.File;
 import java.io.IOException;
+import java.nio.ByteBuffer;
 import java.util.List;
 import org.apache.iceberg.Files;
 import org.apache.iceberg.Schema;
@@ -37,6 +38,8 @@ import org.apache.iceberg.io.FileAppender;
 import org.apache.iceberg.relocated.com.google.common.collect.Lists;
 import org.apache.iceberg.types.Type;
 import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.ByteBuffers;
+import org.apache.iceberg.util.DateTimeUtil;
 import org.junit.Assert;
 import org.junit.Rule;
 import org.junit.Test;
@@ -133,7 +136,7 @@ public class TestReadDefaultValues {
       for (Record record : generatedRecords) {
         Record expectedRecord = GenericRecord.create(readerSchema);
         expectedRecord.set(0, record.get(0));
-        expectedRecord.set(1, defaultValue);
+        expectedRecord.set(1, internalToGeneric(type, defaultValue));
         expected.add(expectedRecord);
       }
 
@@ -166,6 +169,26 @@ public class TestReadDefaultValues {
     }
   }
 
+  static Object internalToGeneric(Type type, Object value) {
+    switch (type.typeId()) {
+      case DATE:
+        return DateTimeUtil.dateFromDays((int) value);
+      case TIME:
+        return DateTimeUtil.timeFromMicros((long) value);
+      case TIMESTAMP:
+        Types.TimestampType timestampType = (Types.TimestampType) type;
+        if (timestampType.shouldAdjustToUTC()) {
+          return DateTimeUtil.timestamptzFromMicros((long) value);
+        } else {
+          return DateTimeUtil.timestampFromMicros((long) value);
+        }
+      case FIXED:
+        return ByteBuffers.toByteArray((ByteBuffer) value);
+    }
+
+    return value;
+  }
+
   // TODO: This class should be removed once NestedField.optional() that takes initialDefault and
   // writeDefault becomes public. It is intentionally package private to avoid exposing it in the
   // public API as of now
```

The tests still fail, but now they show the real problem: default values are not being converted to match the object model. For partition constants, that is done in `GenericReader`:

```java
    Map<Integer, ?> partition =
        PartitionUtil.constantsMap(task, IdentityPartitionConverters::convertConstant);
```

We will need to do something similar for default values. I think what makes the most sense is to add a conversion function to `StructReader` that needs to be implemented by sub-classes. That can call the existing conversion functions, like `IdentityPartitionConverters.convertConstant` or similar for Spark. That particular implementation will need to be updated to convert `ByteBuffer` to `byte[]`, since it currently assumes values will be Avro's `GenericData.Fixed`.

Reply:
That is some deep analysis with 3 levels of conversion to get it right :) I have directly called `IdentityPartitionConverters.convertConstant` from `StructReader` constructor. Ideally, `IdentityPartitionConverters` name should be renamed in a future release since it is no longer specific to partitions.